src/astarte_message_hub.rs

```diff
@@ -71,6 +85,182 @@ where
             astarte_handler,
         }
     }
+
+    /// Create a [tower] layer to intercept the Node ID in the gRPC requests
+    pub fn make_interceptor_layer(&self) -> NodeIdInterceptorLayer {
+        NodeIdInterceptorLayer::new(Arc::clone(&self.nodes))
+    }
+}
+
+/// Errors related to the interception of the NodeId into gRPC request
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+pub enum InterceptorError {
+    /// Node ID not specified in the gRPC request
+    AbsentId,
+    /// Invalid Uuid bytes value for Node ID, {0}
+    Uuid(#[from] uuid::Error),
+    /// Invalid Uuid str value for Node ID, {0}
+    ParseStr(#[from] tonic::metadata::errors::ToStrError),
+    /// The Node ID does not exist
+    IdNotFound,
+    /// Couldn't decode metadata binary value
+    MetadataBytes(#[from] InvalidMetadataValueBytes),
+    /// Error returned by the inner service
+    #[error(transparent)]
+    Inner(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+}
```

Comment:
I would add a trait to convert this in a status code returnable by the Service, so we can return UNAUTHENTICATED for AbsentId and IdNotFound, and INVALID_ARGUMENT for the others.

To return this error you can look at the interceptor service, and even use their structs like the Response and future

Reply:
So, instead of returning an InterfceptorError you would like the service to return a `tonic::Status` code?